### PR TITLE
feat: added support for both click event and dynamic remarketing events

### PR DESCRIPTION
**Pull request by mihir-4116**
## PR Description

Please include a summary of the change along with the relevant motivation and context.

## Notion ticket

Ticket link

## Screenshots

Please add screenshots for any new features or UI bug fixes for the following browsers -

- Chrome
  >
- Firefox
  >
- Safari
  >

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: src/integrations/GoogleAds/browser.js [hunk 1]
```diff
@@ -175,6 +172,36 @@ class GoogleAds {
     return conversionData;
   }
 
+  sendEvent(type, eventName) {
```
Comment (utsabc) [inline src/integrations/GoogleAds/browser.js:175]:
```suggestion
  shouldSendEvent(type, eventName) {
```
Function name is not giving its full intent, also check if you can reduce the cognitive complexity of this function  there are too many nested conditions which makes it difficult to read

File: src/integrations/GoogleAds/browser.js [hunk 2]
```diff
@@ -70,45 +78,41 @@ class GoogleAds {
...
+      const { conversionLabel } = conversionData;
+      const { eventName } = conversionData;
+      const sendToValue = `${this.conversionId}/${conversionLabel}`;
+      let properties = {};
+      if (rudderElement.message.properties) {
+        properties.value = rudderElement.message.properties.revenue;
+        properties.currency = rudderElement.message.properties.currency;
+        properties.transaction_id = rudderElement.message.properties.order_id;
       }
+      properties.send_to = sendToValue;
```
Comment (utsabc) [inline src/integrations/GoogleAds/browser.js:93]:
```suggestion
      const properties = {
         value : rudderElement.message?.properties?.revenue;
         currency : rudderElement.message?.properties?.currency;
         transaction_id : rudderElement.message?.properties?.order_id;
         send_to : sendToValue;
      }

```

File: src/integrations/GoogleAds/browser.js [hunk 3]
```diff
@@ -70,45 +78,41 @@ class GoogleAds {
...
+      properties = removeUndefinedAndNullValues(properties);
+      window.gtag('event', eventName, properties);
+    }
+
+    if (!event) {
+      logger.error('Event name not present');
+      return;
+    }
```
Comment (utsabc) [inline src/integrations/GoogleAds/browser.js:101]: Shouldn't this be checked before  L82?

File: src/integrations/GoogleAds/browser.js [hunk 4]
```diff
@@ -124,33 +128,26 @@ class GoogleAds {
...
+      });
+    }
 
-      if (rudderElement.message.properties) {
-        payload = rudderElement.message.properties;
-      }
+    if (!name) {
+      logger.error('Event name not present');
+      return;
+    }
```
Comment (utsabc) [inline src/integrations/GoogleAds/browser.js:144]: Same as above shouldn't this check be happening on L130?

**Review by utsabc — CHANGES_REQUESTED** [review]

**Comment by utsabc** [comment]
Also please add unit test case for this integration as there isn't any

File: src/integrations/GoogleAds/browser.js (hunk 3 above)
Comment (mihir-4116) [reply src/integrations/GoogleAds/browser.js:101]: for conversion events if event name is not there then we are taking default event name which is `Viewed a Page` from `getConversionData` but for this also `defaultPageConversion` label is required from config. that's why we haven't added condition there

File: src/integrations/GoogleAds/browser.js (hunk 4 above)
Comment (mihir-4116) [reply src/integrations/GoogleAds/browser.js:144]: can you check above comment

File: src/integrations/GoogleAds/browser.js (hunk 3 above)
Comment (utsabc) [reply src/integrations/GoogleAds/browser.js:101]: Yeah but can we add some context (js docs) on the `getConversionData` for this?

Comment (mihir-4116) [reply src/integrations/GoogleAds/browser.js:101]: you mean add comments that if event name is not there then we are taking default name from `getConversionData` function?

Comment (utsabc) [reply src/integrations/GoogleAds/browser.js:101]: Yes add docs to function as intent is not pretty clear on it

File: src/integrations/GoogleAds/browser.js (hunk 2 above)
Comment (mihir-4116) [reply src/integrations/GoogleAds/browser.js]: updated

File: src/integrations/GoogleAds/browser.js (hunk 1 above)
Comment (mihir-4116) [reply src/integrations/GoogleAds/browser.js]: updated

File: .size-limit.js [hunk 5]
```diff
@@ -31,7 +31,7 @@ module.exports = [
     name: 'Core - NPM',
     path: 'dist/npm-lib/index.js',
     gzip: true,
-    limit: '37 kB',
+    limit: '38 kB',
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply .size-limit.js:34]: reduced 0.5 kb size

File: .size-limit.js [hunk 6]
```diff
@@ -19,7 +19,7 @@ module.exports = [
     name: 'All Integrations - CDN',
     path: 'dist/legacy/js-integrations/*.min.js',
     gzip: true,
-    limit: '670 kB',
+    limit: '671 kB',
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply .size-limit.js]: reduced 0.5 kb size

File: __tests__/integrations/GoogleAds/GoogleAds.test.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    googleAds = new GoogleAds({ conversionID: 'AW-11071053757' }, {});
+    googleAds.init();
+    expect(typeof window.gtag).toBe('function');
+  });
+});
+
+describe(googleAdsTrack, () => {
+  let googleAds;
+  beforeEach(() => {
+    googleAds = new GoogleAds(googleAdsConfigs[0], {});
```
Comment (utsabc) [inline __tests__/integrations/GoogleAds/GoogleAds.test.js:34]: Shouldn't we be passing mocked analytics object as second argument ?
ref : https://github.com/rudderlabs/rudder-sdk-js/blob/8f903afcd3a6d5f277fa63c9c24929ce93602456/__tests__/integrations/FacebookPixel/FacebookPixel.test.js#L22

**Comment by utsabc** [comment]
Please use `test` rather than `it` as its the convention in the repo

File: src/integrations/GoogleAds/browser.js [hunk 8]
```diff
@@ -6,24 +6,31 @@ import {
...
     this.pageLoadConversions = config.pageLoadConversions;
     this.clickEventConversions = config.clickEventConversions;
     this.defaultPageConversion = config.defaultPageConversion;
-    this.dynamicRemarketing = config.dynamicRemarketing;
     this.sendPageView = config.sendPageView || true;
     this.conversionLinker = config.conversionLinker || true;
     this.disableAdPersonalization = config.disableAdPersonalization || false;
-    this.name = NAME;
+    this.trackConversions = config.trackConversions;
+    this.trackDynamicRemarketing = config.trackDynamicRemarketing;
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/integrations/GoogleAds/browser.js:26]: this values will always come from config, in config we have added default value - `false` for this

File: __tests__/integrations/GoogleAds/GoogleAds.test.js (hunk 7 above)
Comment (mihir-4116) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js:34]: basically we are passing destination config as first argument and we don't need mockAnalytics here so we are passing empty object

File: __tests__/integrations/GoogleAds/GoogleAds.test.js [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+  productAdded,
+  orderCompleted,
+  googleAdsTrack,
+  googleAdsConfigs,
+  trackCallPayload,
+  mockConversionId,
+  mockEventTypeConversions,
+} from './__mocks__/data';
+
+beforeAll(() => {});
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js]: updated

File: __tests__/integrations/GoogleAds/GoogleAds.test.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+          quantity: 2,
+          currency: 'GBP',
+          typeOfProduct: 'Food',
+        },
+      ],
+    });
+  });
+});
+
+describe(googleAdsTrack, () => {
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js]: updated

File: __tests__/integrations/GoogleAds/GoogleAds.test.js [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+  let googleAds;
+
+  test('Testing init call of Google Ads with ConversionId', () => {
+    googleAds = new GoogleAds({ conversionID: 'AW-11071053757' }, {});
+    googleAds.init();
+    expect(typeof window.gtag).toBe('function');
+  });
+});
+
+describe(googleAdsTrack, () => {
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js]: updated

File: __tests__/integrations/GoogleAds/GoogleAds.test.js [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+          quantity: 2,
+          currency: 'GBP',
+          typeOfProduct: 'Food',
+        },
+      ],
+    });
+  });
+});
+
+describe(googleAdsTrack, () => {
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js]: updated

File: __tests__/integrations/GoogleAds/GoogleAds.test.js [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+    expect(window.gtag.mock.calls[0][1]).toEqual('Order Completed');
+    expect(window.gtag.mock.calls[0][2]).toEqual({
+      currency: 'IND',
+      send_to: `${mockConversionId}/TCBjCKjCs4gYEIXBi58p`,
+      transaction_id: mockOrderId,
+    });
+  });
+});
+
+describe(googleAdsTrack, () => {
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js]: updated

File: src/integrations/GoogleAds/browser.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (utsabc) [reply src/integrations/GoogleAds/browser.js:26]: Still lets add defaults here

File: __tests__/integrations/GoogleAds/GoogleAds.test.js (hunk 7 above)
Comment (utsabc) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js:31]: What about the scenarios where we are testing actual `track` events

**Review by utsabc — APPROVED** [review]

File: .size-limit.js [hunk 14]
```diff
@@ -19,7 +19,7 @@ module.exports = [
     name: 'All Integrations - CDN',
     path: 'dist/legacy/js-integrations/*.min.js',
     gzip: true,
-    limit: '670 kB',
+    limit: '670.5 kB',
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .size-limit.js:22]: `limit: '670.5 kB'` is the gzipped value whereas GoogleAds.min.js is just the minified one.

[the remark this answers is not in the document]
Comment (utsabc) [reply .size-limit.js:22]: is there a way we can calculate exact increase in gzipped value?

File: __tests__/integrations/GoogleAds/GoogleAds.test.js (hunk 7 above)
Comment (mihir-4116) [reply __tests__/integrations/GoogleAds/GoogleAds.test.js:31]: this test cases are for actual track events only

File: src/integrations/GoogleAds/browser.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/integrations/GoogleAds/browser.js:26]: ok

**Review by utsabc — DISMISSED** [review]

File: src/integrations/GoogleAds/utils.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  if (trackEvents && enableFiltering) {
+    const eventNames = [];
+    events.forEach((event) => {
+      if (event.eventName !== '') {
+        eventNames.push(event.eventName);
+      }
+    });
+    return eventNames.includes(eventName.trim());
```
Comment (ItsSudip) [inline src/integrations/GoogleAds/utils.js:28]:
```suggestion
    events.forEach((event) => {
      if(event.eventName === eventName.trim())
        return true;
    });
```

Comment (mihir-4116) [reply src/integrations/GoogleAds/utils.js:28]: @ItsSudip we don't return from forEach loop generally, we will process inside loop and then return final value after loop

File: .size-limit.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .size-limit.js:22]: For a particular destination? yes, we can do that. Use this command: `npm run build:integration:cli --intg=GoogleAds`
You will see something like this:

Screenshot 2023-01-20 at 11 53 39 AM

[the remark this answers is not in the document]
Comment (utsabc) [reply .size-limit.js:22]: Got it cc @mihir-4116

**Review by utsabc — DISMISSED** [review]

File: __tests__/integrations/GoogleAds/GoogleAds.test.js [hunk 16]
```diff
@@ -0,0 +1,196 @@
+import GoogleAds from '../../../src/integrations/GoogleAds/browser';
```
Comment (MoumitaM) [inline __tests__/integrations/GoogleAds/GoogleAds.test.js:1]: Update the file name to browser.test.js

**Review by MoumitaM — APPROVED** [review]

**Review by MoumitaM — APPROVED** [review]

File: src/integrations/GoogleAds/utils.js [hunk 17]
```diff
@@ -31,6 +33,60 @@ function shouldSendEvent(eventName, trackEvents, enableFiltering, events) {
   return false;
 }
 
+/**
+ * returns weather to send conversion event or not
+ * @param {*} eventName
+ * @param {*} trackConversions
+ * @param {*} enableConversionEventsFiltering
+ * @param {*} eventsToTrackConversions
+ * @param {*} dynamicRemarketing
```
Comment (utsabc) [inline src/integrations/GoogleAds/utils.js:42]: can you write short descriptors for each of these params

**Review by utsabc — APPROVED** [review]

File: sanity-suite/src/testBook/TestBook.js [hunk 18]
```diff
@@ -76,17 +76,25 @@ class TestBook {
                           <pre>${JSON.stringify(testCase.inputData, undefined, 2)}</pre>
```
Comment (MoumitaM) [inline sanity-suite/src/testBook/TestBook.js:76]: Except for linting, there is no other changes in this file. Can you revert these changes?

**Review by ItsSudip — DISMISSED** [review]

**Review by ItsSudip — DISMISSED** [review]

**Review by utsabc — DISMISSED** [review]

**Review by MoumitaM — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Review by MoumitaM — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

File: src/integrations/GoogleAds/browser.js [hunk 19]
```diff
@@ -124,64 +165,51 @@ class GoogleAds {
...
-      if (!event) {
-        logger.error('Event name not present');
-        return;
-      }
-
-      let payload = {};
-      const sendToValue = this.conversionId;
+    const { name } = rudderElement.message;
+    const conversionData = getConversionData(
+      this.clickEventConversions,
```
Comment (yashasvibajpai) [inline src/integrations/GoogleAds/browser.js:170]: should have been pageMapping
